src/libsystemd-network/ndisc-router.c

```diff
@@ -715,3 +715,29 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
         *ret_sec = be32toh(*(uint32_t*) (ri + 4));
         return 0;
 }
+
+int sd_ndisc_router_captive_portal_get_uri(sd_ndisc_router *rt, const char **uri, size_t *size) {
+        int r;
+        const char *nd_opt_captive_portal;
+        size_t length;
+
+        assert_return(rt, -EINVAL);
+        assert_return(uri, -EINVAL);
+
+        r = sd_ndisc_router_option_is_type(rt, SD_NDISC_OPTION_CAPTIVE_PORTAL);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return -EMEDIUMTYPE;
+
+        r = sd_ndisc_router_option_get_raw(rt, (void *)&nd_opt_captive_portal, &length);
+        if (r < 0)
+                return r;
+        if (length < 2)
+                return -EBADMSG;
```

Comment:
From the rfc https://www.rfc-editor.org/rfc/rfc8910.html#section-2.3-3.1.1.6
> The URI for the captive portal API endpoint to which the user should connect. This MUST be padded with NUL (0x00) to make the total option length (including the Type and Length fields) a multiple of 8 bytes.

So, maybe better to check the length in more detail?

Reply:
It is guaranteed here. The length field in the RA is in units of 8 octets and the  `length` parameter returned in `sd_ndisc_router_option_get_raw()` is exactly the length field \* 8, so length is always a multiple of 8.

The only question is whether or not the client padded the remaining 0-7 bytes with NULs. I use `strnlen` below to find the exact length of the uri without trailing NULs. Should I check something more?

Reply:
Hm I should probably check that the strnlen result doesn't truncate the option before the last 8 bytes.